diff --git a/CMakeLists.txt b/CMakeLists.txt
index e404c5a..1520a3f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,11 +16,7 @@ if(UNIX AND NOT APPLE AND NOT CYGWIN)
     set(ICONDIR "${SHAREDIR}/pixmaps" CACHE STRING "Install location of the icon file")
     set(DESKTOPDIR "${SHAREDIR}/applications" CACHE STRING "Install location of the desktop shortcut")
 endif()
-if(CMAKE_BUILD_TYPE MATCHES Release)
-    set(DATADIR "${SHAREDIR}/openblok" CACHE STRING "Install location of the data files (release)")
-else()
-    set(DATADIR "data" CACHE STRING "Install location of the data files (debug)")
-endif()
+set(DATADIR "data" CACHE STRING "Install location of the data files (debug)")
 
 
 # Global compiler settings
@@ -45,6 +41,25 @@ if(BUILD_TEST_COVERAGE)
 endif()
 
 
+# Emscripten
+if(EMSCRIPTEN)
+    set(CMAKE_EXECUTABLE_SUFFIX ".bc")
+    set(EMOPTIONS "${EMOPTIONS} -std=c++14") # bug?
+    set(EMOPTIONS "${EMOPTIONS} -O3")
+    # set(EMOPTIONS "${EMOPTIONS} -O0 -g4 -s ASSERTIONS=2 -s DEMANGLE_SUPPORT=1") # for debugging
+    set(EMOPTIONS "${EMOPTIONS} -s USE_SDL=2")
+    set(EMOPTIONS "${EMOPTIONS} -s USE_SDL_IMAGE=2 -s SDL2_IMAGE_FORMATS='[\"png\"]'")
+    set(EMOPTIONS "${EMOPTIONS} -s USE_SDL_TTF=2")
+    set(EMOPTIONS "${EMOPTIONS} -s USE_OGG=1 -s USE_VORBIS=1")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${EMOPTIONS}")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${EMOPTIONS}")
+    set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} --preload-file ${CMAKE_SOURCE_DIR}/data@data")
+
+    # SDL2_mixer
+    include_directories(external/SDL2_mixer/installdir/include)
+endif()
+
+
 # Dependencies: SDL2
 set(SDL2PP_CXXSTD "c++14" CACHE STRING "libSDL2pp C++ standard")
 set(SDL2PP_WITH_IMAGE ON)
diff --git a/external/libSDL2pp/CMakeLists.txt b/external/libSDL2pp/CMakeLists.txt
index aaa6d90..d08669b 100644
--- a/external/libSDL2pp/CMakeLists.txt
+++ b/external/libSDL2pp/CMakeLists.txt
@@ -26,7 +26,7 @@ ELSE(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
 ENDIF(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
 
 # depends
-FIND_PACKAGE(SDL2 REQUIRED)
+# FIND_PACKAGE(SDL2 REQUIRED)
 SET(SDL2_ALL_INCLUDE_DIRS ${SDL2_INCLUDE_DIR})
 SET(SDL2_ALL_LIBRARIES ${SDL2_LIBRARY})
 SET(SDL2_ALL_PKGCONFIG_MODULES sdl2)
@@ -40,7 +40,7 @@ IF(MINGW)
 ENDIF(MINGW)
 
 IF(SDL2PP_WITH_IMAGE)
-	FIND_PACKAGE(SDL2_image REQUIRED)
+	# FIND_PACKAGE(SDL2_image REQUIRED)
 	SET(SDL2_ALL_INCLUDE_DIRS ${SDL2_ALL_INCLUDE_DIRS} ${SDL2_IMAGE_INCLUDE_DIR})
 	SET(SDL2_ALL_LIBRARIES ${SDL2_ALL_LIBRARIES} ${SDL2_IMAGE_LIBRARY})
 	SET(SDL2_ALL_PKGCONFIG_MODULES "${SDL2_ALL_PKGCONFIG_MODULES} SDL2_image")
@@ -49,7 +49,7 @@ ELSE(SDL2PP_WITH_IMAGE)
 ENDIF(SDL2PP_WITH_IMAGE)
 
 IF(SDL2PP_WITH_TTF)
-	FIND_PACKAGE(SDL2_ttf REQUIRED)
+	# FIND_PACKAGE(SDL2_ttf REQUIRED)
 	SET(SDL2_ALL_INCLUDE_DIRS ${SDL2_ALL_INCLUDE_DIRS} ${SDL2_TTF_INCLUDE_DIR})
 	SET(SDL2_ALL_LIBRARIES ${SDL2_ALL_LIBRARIES} ${SDL2_TTF_LIBRARY})
 	SET(SDL2_ALL_PKGCONFIG_MODULES "${SDL2_ALL_PKGCONFIG_MODULES} SDL2_ttf")
@@ -58,7 +58,7 @@ ELSE(SDL2PP_WITH_TTF)
 ENDIF(SDL2PP_WITH_TTF)
 
 IF(SDL2PP_WITH_MIXER)
-	FIND_PACKAGE(SDL2_mixer REQUIRED)
+	# FIND_PACKAGE(SDL2_mixer REQUIRED)
 	SET(SDL2_ALL_INCLUDE_DIRS ${SDL2_ALL_INCLUDE_DIRS} ${SDL2_MIXER_INCLUDE_DIR})
 	SET(SDL2_ALL_LIBRARIES ${SDL2_ALL_LIBRARIES} ${SDL2_MIXER_LIBRARY})
 	SET(SDL2_ALL_PKGCONFIG_MODULES "${SDL2_ALL_PKGCONFIG_MODULES} SDL2_mixer")
@@ -94,7 +94,7 @@ ELSE(MSVC)
 	#APPEND_CXX_FLAG_IF_SUPPORTED(-Wno-padded SDL2PP_WARNINGS)
 ENDIF(MSVC)
 
-LIST(REMOVE_DUPLICATES SDL2_ALL_INCLUDE_DIRS)
+# LIST(REMOVE_DUPLICATES SDL2_ALL_INCLUDE_DIRS)
 
 INCLUDE_DIRECTORIES(BEFORE ${PROJECT_SOURCE_DIR})
 INCLUDE_DIRECTORIES(SYSTEM ${SDL2_ALL_INCLUDE_DIRS})
diff --git a/external/libSDL2pp/SDL2pp/Font.cc b/external/libSDL2pp/SDL2pp/Font.cc
index 5d2a181..705f9fa 100644
--- a/external/libSDL2pp/SDL2pp/Font.cc
+++ b/external/libSDL2pp/SDL2pp/Font.cc
@@ -21,7 +21,7 @@
 
 #include <vector>
 
-#include <SDL2/SDL_ttf.h>
+#include <SDL_ttf.h>
 
 #include <SDL2pp/Font.hh>
 #include <SDL2pp/RWops.hh>
diff --git a/external/libSDL2pp/SDL2pp/Font.hh b/external/libSDL2pp/SDL2pp/Font.hh
index 32abb8f..44587e1 100644
--- a/external/libSDL2pp/SDL2pp/Font.hh
+++ b/external/libSDL2pp/SDL2pp/Font.hh
@@ -24,7 +24,7 @@
 
 #include <string>
 
-#include <SDL2/SDL_ttf.h>
+#include <SDL_ttf.h>
 
 #include <SDL2pp/Optional.hh>
 #include <SDL2pp/Point.hh>
diff --git a/external/libSDL2pp/SDL2pp/SDLTTF.cc b/external/libSDL2pp/SDL2pp/SDLTTF.cc
index 0eb187d..dfc067b 100644
--- a/external/libSDL2pp/SDL2pp/SDLTTF.cc
+++ b/external/libSDL2pp/SDL2pp/SDLTTF.cc
@@ -19,7 +19,7 @@
   3. This notice may not be removed or altered from any source distribution.
 */
 
-#include <SDL2/SDL_ttf.h>
+#include <SDL_ttf.h>
 
 #include <SDL2pp/SDLTTF.hh>
 #include <SDL2pp/Exception.hh>
diff --git a/src/main.cpp b/src/main.cpp
index bc5b424..319cfdf 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -22,15 +22,17 @@
 #include "system/Log.h"
 #include "system/Paths.h"
 
-#include <chrono>
 #include <memory>
-#include <thread>
 #include <assert.h>
+#include <emscripten.h>
 
 
 const std::string LOG_MAIN = "main";
 const std::string LOG_HELP = "help";
 
+AppContext app;
+void runNextFrame();
+
 int main(int argc, const char** argv)
 {
     Log::info(LOG_MAIN) << "OpenBlok v0.1.0 by Mátyás Mustoha\n";
@@ -60,7 +62,6 @@ int main(int argc, const char** argv)
     }
 
 
-    AppContext app;
     if (!app.init())
         return 1;
 
@@ -72,31 +73,24 @@ int main(int argc, const char** argv)
     }
 
 
-    auto frame_starttime = std::chrono::steady_clock::now();
-    auto frame_endtime = frame_starttime + Timing::frame_duration;
-
-    while (!app.window().quitRequested()) {
-        try {
-            auto events = app.window().collectEvents();
+    emscripten_set_main_loop(runNextFrame, 0, 1);
 
-            assert(app.states().size());
-            app.states().top()->update(events, app);
-            if (app.states().empty())
-                break;
-
-            app.states().top()->draw(app.gcx());
-            app.gcx().render();
-        }
-        catch (const std::exception& err) {
-            app.window().showErrorMessage(err.what());
-            return 1;
-        }
+    return 0;
+}
 
-        // frame rate limiting
-        std::this_thread::sleep_until(frame_endtime);
-        frame_starttime = std::chrono::steady_clock::now();
-        frame_endtime = frame_starttime + Timing::frame_duration;
+void runNextFrame()
+{
+    try {
+        auto events = app.window().collectEvents();
+        assert(app.states().size());
+        app.states().top()->update(events, app);
+        if (app.states().empty())
+            return;
+
+        app.states().top()->draw(app.gcx());
+        app.gcx().render();
+    }
+    catch (const std::exception& err) {
+        app.window().showErrorMessage(err.what());
     }
-
-    return 0;
 }
diff --git a/src/system/sdl/SDLGraphicsContext.cpp b/src/system/sdl/SDLGraphicsContext.cpp
index 64821fd..9cdbe97 100644
--- a/src/system/sdl/SDLGraphicsContext.cpp
+++ b/src/system/sdl/SDLGraphicsContext.cpp
@@ -51,7 +51,7 @@ void SDLGraphicsContext::render()
     renderer.Present();
     on_render_callback();
 
-    renderer.Clear();
+    // renderer.Clear();
 }
 
 unsigned short SDLGraphicsContext::screenWidth() const
